File: drug-discovery-ui/src/services/reactions.ts
export type ReactionRequest = {
  reactantA: string
  reactantB: string
  conditions: { temperature?: string; pressure?: string; catalyst?: string; solvent?: string }
}

export type ReactionResponse = {
  equation: string
  products: { name: string; confidence: number; byproduct: boolean }[]
  mechanismSteps: string[]
  yieldPercent: number
  energyChange: 'exothermic' | 'endothermic' | 'neutral'
  heuristic?: boolean
  error?: string
}

export async function predictReaction(input: ReactionRequest): Promise<ReactionResponse> {
  const resp = await fetch('/api/v1/reactions/predict', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
  if (!resp.ok) throw new Error(`Server error: ${resp.status}`)
  return (await resp.json()) as ReactionResponse
}

File: drug-discovery-ui/backend/app/api/models/reactions.py
from typing import List, Optional, Literal
from pydantic import BaseModel, Field

EnergyChange = Literal['exothermic', 'endothermic', 'neutral']

class ReactionConditions(BaseModel):
    temperature: Optional[str] = Field(None, description='e.g., 25 C')
    pressure: Optional[str] = Field(None, description='e.g., 1 atm')
    catalyst: Optional[str] = None
    solvent: Optional[str] = None

class ReactionRequest(BaseModel):
    reactantA: str = Field(..., min_length=1)
    reactantB: str = Field(..., min_length=1)
    conditions: ReactionConditions = ReactionConditions()

class ReactionProduct(BaseModel):
    name: str
    confidence: int = Field(ge=0, le=100)
    byproduct: bool = False

class ReactionResponse(BaseModel):
    equation: str
    products: List[ReactionProduct]
    mechanismSteps: List[str]
    yieldPercent: int = Field(ge=0, le=100)
    energyChange: EnergyChange = 'neutral'
    heuristic: bool = False
    error: Optional[str] = None
